pkg/controller/registry/reconciler/configmap.go

```diff
@@ -513,5 +515,36 @@ func (c *ConfigMapRegistryReconciler) CheckRegistryServer(logger *logrus.Entry,
 	}
 
 	healthy = true
+	if deadPodsDetected, e := detectAndDeleteDeadPods(logger, c.OpClient, pods, source.GetNamespace()); deadPodsDetected {
+		healthy = false
+		err = e
+	}
+	return
+}
+
+// detectAndDeleteDeadPods determines if there are registry client pods that are in the deleted state
+// but have not been removed by GC (eg the node goes down before GC can remove them), and attempts to
+// force delete the pods
+func detectAndDeleteDeadPods(logger *logrus.Entry, client operatorclient.ClientInterface, pods []*corev1.Pod, sourceNamespace string) (foundDeadPod bool, err error) {
+	var forceDeletionErrs []error
+	foundDeadPod = false
+	for _, pod := range pods {
+		if !isPodDead(pod) {
+			logger.WithFields(logrus.Fields{"pod.namespace": sourceNamespace, "pod.name": pod.GetName()}).Debug("pod is alive")
+			continue
+		}
+		foundDeadPod = true
```

Comment:
This is trickier than I originally thought. If we have:
1. at least one alive pod
2. the alive pods are otherwise deemed healthy
3. we successfully delete the dead pods

What should we return from `CheckRegistryServer`? Seems like we could say healthy in that case?

Reply:
Yea this is a bit confusing, I'm thinking if we detect even one dead pod, we just force delete all the pods, and let `EnsureRegistryServer` recreate everything. Otherwise we're in a very non-deterministic state...

Reply:
That seems like that could have significant performance implications. There's a lot of disk/cpu/memory costs that are paid when catalog pods are started, so we should minimized that as much as possible.

I think it is okay to force delete the dead pods, and if there are any alive pods, we just move forward as if those were the only pods that were there to begin with.

Reply:
That's fair too. I was a bit confused about there being multiple pods in the first place, since my impression is that we only create one registry pod per catalog, but turns out we have `pods` mainly because we use the lister to list via label selector. In any case this should return one pod, unless I'm unaware of some feature that allows us to spin up multiple registry pods per catalog. 

Changed it back to just deleting dead pods.

Reply:
There are definitely multiple pods when we're polling. In that case, we spin up a new pod to see what the digest is, and we compare the digests to see if we need to use the new pod (if it has a new digest) or keep using the old pod (when the digests match)

There may also be multiple pods when the pod spec changes (e.g. when the catalog source image is changed).

Reply:
Right, both are transitory states though, and we're in that state because `EnsureRegistryServer` has already been called. But yes I think this is in a good state now.